cnd/src/swap_protocols/ledger_states.rs

```diff
@@ -0,0 +1,126 @@
+use crate::swap_protocols::{
+    rfc003::{create_swap::SwapEvent, LedgerState},
+    state::{Get, Insert, Update},
+    swap_id::SwapId,
+};
+use std::{
+    any::Any,
+    collections::HashMap,
+    ops::{Deref, DerefMut},
+    sync::Mutex,
+};
+
+#[derive(Default, Debug)]
+pub struct AlphaLedgerState {
+    states: Mutex<HashMap<SwapId, Box<dyn Any + Send>>>,
+}
+
+#[derive(Default, Debug)]
+pub struct BetaLedgerState {
+    states: Mutex<HashMap<SwapId, Box<dyn Any + Send>>>,
+}
+
+impl<S, L> Insert<S> for L
+where
+    L: DerefMut<Target = Mutex<HashMap<SwapId, Box<dyn Any + Send>>>> + Send + Sync + 'static,
+    S: Send + 'static,
+{
+    fn insert(&self, key: SwapId, value: S) {
+        let mut states = self.lock().unwrap();
+        states.insert(key, Box::new(value));
+    }
+}
+
+impl<S, L> Get<S> for L
+where
+    L: DerefMut<Target = Mutex<HashMap<SwapId, Box<dyn Any + Send>>>> + Send + Sync + 'static,
+    S: Clone + Send + 'static,
+{
+    fn get(&self, key: &SwapId) -> anyhow::Result<Option<S>> {
+        let states = self.lock().unwrap();
+        match states.get(key) {
+            Some(state) => match state.downcast_ref::<S>() {
+                Some(state) => Ok(Some(state.clone())),
+                None => Err(anyhow::anyhow!("invalid type")),
+            },
+            None => Ok(None),
+        }
+    }
+}
+
+impl<A, H, T, L> Update<LedgerState<A, H, T>, SwapEvent<A, H, T>> for L
+where
+    L: DerefMut<Target = Mutex<HashMap<SwapId, Box<dyn Any + Send>>>> + Send + Sync + 'static,
+    LedgerState<A, H, T>: 'static,
+{
+    fn update(&self, key: &SwapId, event: SwapEvent<A, H, T>) {
```

Comment:
Can we have a comment on this function please, the return value is not immediately obvious.

Reply:
There is no return value.